### DISCUSSION TITLE
Argon2 for local master password hashing (not encryption key derivation)

### DISCUSSION THREAD

**Discussion by junkurihara** [dani-garcia/vaultwarden#2558] [Ideas]
Hello!

First of all, I really appreciate for this great port of Bitwarden in Rust!

I am a bit new to here, but I read the source code and found that (somewhat legacy) PBKDF2 is used for "local master password hashing."

Of course, I really understand that the encryption key is still derived by PBKDF2 due to compatibility with original Bitwarden and for compatibility with various frontends like iOS/Android Apps (see https://community.bitwarden.com/t/switch-to-argon2/350). So I know it is tough to quickly change the algorithm to modern one for the encryption key derivation. On the other hand, I think that in current era, since local password hashing has no compatibility issue with Bitwarden, Argon2 could be better in terms of "modern" cryptography, whereas PBKDF2 is still no problem. 

I actually implemented the Argon2 version of Vaultwarden by myself in my forked repo (https://github.com/junkurihara/vaultwarden), and it works with no problem at this point. It can also compatible with existing users with PBKDF2 password hashing, and they can migrate to Argon2 when they update their master passwords.

How do you think of this idea? I would be really happy if you give me any feedback! If you would like this, I can make a PR, of course. Thanks a lot!

Best regards,
Jun

**Comment by bokkabonga** [comment]
I don´t think that this is something to be implemented here in vaultwarden in order to keep it as close to upstream as possible and reduce the chance of future problems.

According to the bitwarden devs argon2 is still "on our radar screen but not slated for 2022." So i guess it makes sense to wait a bit longer for the official implementation

**Comment by BlackDex** [comment]
I think using something different the PBKDF2 isn't that useful actually, for one, it is only used to verify a password, and not decrypt/encrypt anything. Second, the Bitwarden clients only support PBKDF2 and have there iterations configured for that in the interface (web-vault).

Also, i don't know if it this is any consideration, but PBKDF2 is FIPS-140 compliant, which Argon2 isn't, which could be an issue for some users.

And, as @bokkabonga also mentioned, want to keep as close to Bitwarden as possible, especially regarding the encryption protocols the use and support.

**Comment by d235j** [comment]
With all the discourse around Lastpass and its master passwords being crackable, It looks like upstream is in the process of implementing Argon2: https://github.com/bitwarden/clients/pull/4468 https://github.com/bitwarden/server/pull/2583

Once this is merged, it would be worth revisiting the above changes.

**Reply by BlackDex** [reply]
That will not have any useful effect. Since the articles regarding the iterations also mentions that changing this on the server side would not have any effect on this. Still, i have created a PR #3163 to at least increase the amount of iterations by default. And also that it will update the iterations of the already stored PBKDF values.

Again, it will not add any extra security to the server side implementation if we change this to something else then we already have.

**Reply by BlackDex** [reply]
Just to be clear here btw. Once they have added Argon2 encryption on the client side, we still need to implement this on our server side.
It's probably just a few extra records, but we have to wait until they have finalized everything before we can start implementing this.

**Reply by quexten** [reply]
Ah, didn't see this comment before opening https://github.com/dani-garcia/vaultwarden/discussions/3171 .But yes it's just 2 records in the user table, and a couple of changes for the different requests/responses that include iterations.

**Comment by musicislife08** [comment]
Looks like as of client version 2023.2 the official bitwarden server and clients now support argon2

https://www.ghacks.net/2023/02/20/bitwarden-password-manager-introduces-support-for-argon2-kdf-iterations/

also listed in the official bitwarden documentation as a note

https://bitwarden.com/help/kdf-algorithms/#changing-kdf-algorithm
